server/protocols/jmap/src/main/java/org/apache/james/jmap/mailet/TextCalendarBodyToAttachment.java

```diff
@@ -0,0 +1,104 @@
+
+package org.apache.james.jmap.mailet;
+
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+
+import javax.mail.Header;
+import javax.mail.MessagingException;
+import javax.mail.Multipart;
+import javax.mail.Part;
+import javax.mail.internet.MimeBodyPart;
+import javax.mail.internet.MimeMessage;
+import javax.mail.internet.MimeMultipart;
+
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMailet;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * This mailet converts Content-Type of MimeMessage from text/calendar to mulitpart/mixed
+ *
+ * The BodyPart should be retrieved from content of text/calendar with all the same "Content-*" headers from original message
+ * and those "Content-" header are removed from original message
+ *
+ * <br />
+ * It does not takes any parameter
+ *
+ * Sample configuration:
+ * <p/>
+ * <pre><code>
+ * &lt;mailet match="All" class="TextCalendarBodyToAttachment"/&gt;
+ * </code></pre>
+ *
+ */
+public class TextCalendarBodyToAttachment extends GenericMailet {
+    private static final String TEXT_CALENDAR_TYPE = "text/calendar";
+    private static final String CONTENT_HEADER_PREFIX = "Content-";
+
+    @Override
+    public String getMailetInfo() {
+        return "Moves body part of content type text/calendar to attachment";
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        MimeMessage mimeMessage = mail.getMessage();
+        if (mimeMessage.isMimeType(TEXT_CALENDAR_TYPE)) {
+            processTextBodyAsAttachment(mimeMessage);
+        }
+    }
+
+    @VisibleForTesting
+    void processTextBodyAsAttachment(MimeMessage mimeMessage) throws MessagingException {
+        Multipart multipart = new MimeMultipart();
+        multipart.addBodyPart(createMimeBodyPartFromMimeMessage(mimeMessage));
+
+        mimeMessage.setContent(multipart);
+        mimeMessage.saveChanges();
+    }
+
+    private MimeBodyPart createMimeBodyPartFromMimeMessage(MimeMessage mimeMessage) throws MessagingException {
+        MimeBodyPart fileBody = createMimeBodyPartAndMoveContentHeadersFromMimeMessageToMimePart(mimeMessage);
+
+        fileBody.setDisposition(Part.ATTACHMENT);
+        return fileBody;
+    }
+
+    private MimeBodyPart createMimeBodyPartAndMoveContentHeadersFromMimeMessageToMimePart(MimeMessage mimeMessage) throws MessagingException {
+        MimeBodyPart fileBody = new MimeBodyPart(mimeMessage.getRawInputStream());
+        List<Header> contentHeaders = Collections.list((Enumeration<Header>) mimeMessage.getAllHeaders())
+            .stream()
+            .filter(header -> header.getName().startsWith(CONTENT_HEADER_PREFIX))
+            .collect(Guavate.toImmutableList());
+
+        for (Header header : contentHeaders) {
+            fileBody.setHeader(header.getName(), header.getValue());
+            mimeMessage.removeHeader(header.getName());
```

Comment:
I'm coming late, but this line should be extracted.
It will allow to have the method createMimeBodyPartFromMimeMessage respect its naming and to have no side effect on mimeMessage argument.